x/datadeal/keeper/deal.go

```diff
@@ -0,0 +1,148 @@
+package keeper
+
+import (
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	gogotypes "github.com/gogo/protobuf/types"
+	"github.com/medibloc/panacea-core/v2/x/datadeal/types"
+)
+
+func (k Keeper) CreateDeal(ctx sdk.Context, consumerAddress sdk.AccAddress, msg *types.MsgCreateDeal) (uint64, error) {
+	dealID, err := k.GetNextDealNumberAndIncrement(ctx)
+	if err != nil {
+		return 0, sdkerrors.Wrapf(err, "failed to get next deal num")
+	}
+
+	newDeal := types.NewDeal(dealID, msg)
+
+	coins := sdk.NewCoins(*msg.Budget)
+
+	dealAddress, err := sdk.AccAddressFromBech32(newDeal.Address)
+	if err != nil {
+		return 0, err
+	}
+
+	acc := k.accountKeeper.GetAccount(ctx, dealAddress)
+	if acc != nil {
+		return 0, sdkerrors.Wrapf(types.ErrDealAlreadyExist, "deal %d already exist", dealID)
+	}
+
+	acc = k.accountKeeper.NewAccount(ctx, authtypes.NewModuleAccount(
+		authtypes.NewBaseAccountWithAddress(
+			dealAddress,
+		),
+		"deal"+strconv.FormatUint(newDeal.Id, 10)),
+	)
+	k.accountKeeper.SetAccount(ctx, acc)
+
+	if err = k.bankKeeper.SendCoins(ctx, consumerAddress, dealAddress, coins); err != nil {
+		return 0, sdkerrors.Wrapf(err, "Failed to send coins to deal account")
+	}
+
+	if err = k.SetDeal(ctx, newDeal); err != nil {
+		return 0, err
+	}
+
+	return newDeal.Id, nil
+}
+
+func (k Keeper) SetNextDealNumber(ctx sdk.Context, dealNumber uint64) error {
+	store := ctx.KVStore(k.storeKey)
+	bz, err := k.cdc.MarshalLengthPrefixed(&gogotypes.UInt64Value{Value: dealNumber})
+	if err != nil {
+		return sdkerrors.Wrapf(err, "Failed to set next deal number")
+	}
+	store.Set(types.DealNextNumberKey, bz)
+	return nil
+}
+
+func (k Keeper) GetNextDealNumber(ctx sdk.Context) (uint64, error) {
+	var dealNumber uint64
+	store := ctx.KVStore(k.storeKey)
+
+	if !store.Has(types.DealNextNumberKey) {
+		return 0, types.ErrDealNotInitialized
+	}
+
+	bz := store.Get(types.DealNextNumberKey)
+
+	val := gogotypes.UInt64Value{}
+
+	if err := k.cdc.UnmarshalLengthPrefixed(bz, &val); err != nil {
+		return 0, sdkerrors.Wrapf(err, "Failed to get next deal number")
+	}
+
+	dealNumber = val.GetValue()
+
+	return dealNumber, nil
+}
+
+func (k Keeper) GetNextDealNumberAndIncrement(ctx sdk.Context) (uint64, error) {
```

Comment:
Usually, we name this kind of functions as `Verb1AndVerb2...`. So, we can rename this to `GetAndIncreaseNextDealNumber`.
https://en.wikipedia.org/wiki/Compare-and-swap

Reply:
Fixed in [8077e85](https://github.com/medibloc/panacea-core/pull/525/commits/8077e85ac4a079ae1a55d5c181a0047c1cb22a22)